lib/utils/addSemicolonForEditInfo.mjs

```diff
@@ -0,0 +1,30 @@
+/** @import { Node as PostcssNode } from 'postcss' */
+
+/**
+ * Adjust the text in EditInfo to include a semicolon when needed.
+ *
+ * @param {PostcssNode} node
+ * @param {{range: [number, number], text: string}} fixData
+ * @returns {{range: [number, number], text: string}}
+ */
+export default function addSemicolonForEditInfo(node, fixData) {
+	if (node.type === 'decl') {
+		if (node.parent?.raws.semicolon || node.parent?.last !== node) {
+			return {
+				...fixData,
+				text: `${fixData.text};`,
+			};
+		}
+	}
+
+	if (node.type === 'atrule') {
+		if (!('nodes' in node) && (node.parent?.raws.semicolon || node.parent?.last !== node)) {
```

Comment:
[suggestion] Checking the `nodes` property seems unnecessary because [`AtRules#nodes`](https://postcss.org/api/#atrule-nodes) always exists.

```suggestion
		if (node.parent?.raws.semicolon || node.parent?.last !== node) {
```

Reply:
`nodes` is missing when it is a statement instead of block.
See: https://github.com/postcss/postcss/blob/755f08f8dbdbc001906b10db3b262948f8491d82/lib/stringifier.js#L37-L43

```css
@import "foo.css"; /* a statement */
```

```css
@media screen {} /* a block */
```

Only at rule statements should be handled by `addSemicolonForEditInfo`

Reply:
Oh... that's correct, I didn't realize it 😅 

```js
> postcss.parse('@charset "utf-8"').nodes[0].type
'atrule'
> postcss.parse('@charset "utf-8"').nodes[0].nodes
undefined
```

So, PostCSS probably should fix its inaccurate type of `AtRule` like this:

```diff
- nodes: Container['nodes']
+ nodes?: Container['nodes']
```

https://github.com/postcss/postcss/blob/755f08f8dbdbc001906b10db3b262948f8491d82/lib/at-rule.d.ts#L100

Reply:
Indeed!

I've filed: https://github.com/postcss/postcss/issues/2019

Reply:
Thank you! 😄

Reply:
I just merged #8408 to bump PostCSS to 8.5.3, for which @romainmenke included type fixes. 👏🏼 

@pamelalozano16 Can you please merge the main branch into this branch?

Reply:
Here's the `AtRule#nodes` type on 8.5.3:

https://github.com/postcss/postcss/blob/8.5.3/lib/at-rule.d.ts#L100

```ts
  nodes: Container['nodes'] | undefined
```

Do you think `!('nodes' in node)` still makes sense in this PR? It seems that `'nodes' in node` always returns `true`. 🤔 

If this code checks if the `nodes` property is empty, the following might be more accurate by using the [`isAtRule` type guard](https://github.com/stylelint/stylelint/blob/b7e705d751be28898ab67081a1d2b70ec94669a2/lib/utils/typeGuards.mjs#L24-L26):

```diff
-	if (node.type === 'atrule') {
-		if (!('nodes' in node) && (parent.raws.semicolon || parent.last !== node)) {
+	if (isAtRule(node)) {
+		if (!node.nodes && (parent.raws.semicolon || parent.last !== node)) {
```

Any thoughts?

Reply:
Looks good, done!